implement Replica::pending_tasks

Scandiravian opened the pull request:
This is a first attempt at #452. 

I've implemented two different approaches in the storage layer and I'm looking for feedback on whether either of these, or a third approach, would be most optimal. 

Currently the `taskdb` layer use the more "naive" approach, which follows the original c++ code more closely. 

The alternative approach would use the `get_tasks_by_status` to query for `pending` tasks. I am however not sure whether there could potentially be an inconsistency between the `working_set` and the sqlite database, which would make that solution unfeasible.

Commit:
add Replica::pending_tasks

djmitche commented:
Nice work! I think it would be unusual to want to get all completed or deleted tasks (and those cases would be covered by `all_tasks()`).

And, the set of all pending tasks and the working set can be slightly different in ways that might be significant. I think the best approach here would be joining with the working set to get the working-set tasks.

Scandiravian commented:
> Nice work! I think it would be unusual to want to get all completed or deleted tasks (and those cases would be covered by `all_tasks()`).
> 
> And, the set of all pending tasks and the working set can be slightly different in ways that might be significant. I think the best approach here would be joining with the working set to get the working-set tasks.

Thanks for the feedback. I've updated the PR so it hopefully reflect your suggestions. I moved some code from `pending_tasks` to `pending_tasks_data` when implementing the latter, but the functionality should be the same.

I have also fixed the CI failures, so I think this is ready for review. Let me know if you'd like any other changes to the code

djmitche commented:
> > Nice work! I think it would be unusual to want to get all completed or deleted tasks (and those cases would be covered by `all_tasks()`).
> > And, the set of all pending tasks and the working set can be slightly different in ways that might be significant. I think the best approach here would be joining with the working set to get the working-set tasks.
> 
> Thanks for the feedback. I've updated the PR so it hopefully reflect your suggestions. I moved some code from `pending_tasks` to `pending_tasks_data` when implementing the latter, but the functionality should be the same.
> 
> I have also fixed the CI failures, so I think this is ready for review. Let me know if you'd like any other changes to the code

This implementation looks solid and well-tested!

However, this still requires two separate queries to get the pending tasks: first to get the list of UUIDs in the pending set, and second to fetch each of those tasks. This involves sending that list of UUIDs from the DB to Rust and back. Could a single `get_pending_tasks` method do this in one query, with something like `SELECT tasks.* from tasks join working_set on..`?

Commit:
get pending set in one query by joining tasks on working_set

Scandiravian commented:
That's really good feedback. I hadn't realised that the working set already exists in the database, thanks :smile: 

I'm guessing that there's somewhere else in the code that ensures only tasks with the "pending" status are added to the working_set?

I've added a new commit that changes the implementation to using a join to do it in a single query. Let me know if you would like any other changes. I'm personally not entirely happy with the `inmemory` implementation of `get_pending_tasks` as it's a bit too nested/complicated making it difficult to read

If you're happy with how the implementation is looking I'm going to squash the changes into a single commit before merging and give the code another pass to see if I can improve the readability+improve the doc strings

djmitche commented:
> I'm guessing that there's somewhere else in the code that ensures only tasks with the "pending" status are added to the working_set?

Yes, modifications to the `status` property trigger adjustments of the working set.
 
> I've added a new commit that changes the implementation to using a join to do it in a single query. Let me know if you would like any other changes. I'm personally not entirely happy with the `inmemory` implementation of `get_pending_tasks` as it's a bit too nested/complicated making it difficult to read
> 
> If you're happy with how the implementation is looking I'm going to squash the changes into a single commit before merging and give the code another pass to see if I can improve the readability+improve the doc strings

I'll have a look shortly!

djmitche commented on taskchampion/src/replica.rs:
```diff
@@ -638,6 +661,30 @@ mod tests {
+
+        rep.commit_operations(ops).unwrap();
+
+        let all_tasks = rep.pending_tasks().unwrap();
```
Just for accuracy's sake, please call tihs `pending_tasks`..

Scandiravian replied:
Thanks! Wouldn't have caught this myself

djmitche commented on taskchampion/src/storage/inmemory.rs:
```diff
@@ -49,6 +49,24 @@ impl<'t> StorageTxn for Txn<'t> {
+                })
+                .flatten()
+            })
+            .collect::<Vec<_>>();
```
This is nice! It took me a bit to realize that `.get(..)` also returned an option, resulting in `Option<Option<(uuid, TaskMap)>>` and requiring the `.flatten()`.

Scandiravian replied:
This is the part that I dislike because it's not readable enough

I'm not smart enough to figure out how to make it cleaner, so I've added a comment about what is going on to make it a bit more clear

djmitche replied:
That's probably good. I think this is actually pretty idiomatic Rust, but it does require a lot of context about the various types involved to understand it. The comment addresses that perfectly. 💯

djmitche commented on taskchampion/src/storage/test.rs:
```diff
@@ -267,6 +272,54 @@ pub(super) fn delete_task_exists(mut storage: impl Storage) -> Result<()> {
+        tasks.sort_by(|a, b| a.0.cmp(&b.0));
+
+        assert_eq!(tasks, exp);
+    }
```
I modified this test as follows, and it fails for sqlite. I think this is because of the `LEFT JOIN` - a task which matches no rows in the working set is included in the result set. I think the query should just be `JOIN` (and I think I led you astray in an earlier comment -- sorry!)

```diff
diff --git taskchampion/src/storage/test.rs taskchampion/src/storage/test.rs
index 390bf2f6c..9fe503fd3 100644
--- taskchampion/src/storage/test.rs
+++ taskchampion/src/storage/test.rs
@@ -268,38 +268,41 @@ pub(super) fn delete_task_exists(mut storage: impl Storage) -> Result<()> {
     {
         let mut txn = storage.txn()?;
         assert!(txn.delete_task(uuid)?);
     }
     Ok(())
 }
 
 pub(super) fn get_pending_tasks(mut storage: impl Storage) -> Result<()> {
-    let uuid1 = Uuid::new_v4();
-    let uuid2 = Uuid::new_v4();
+    let uuid1 = Uuid::new_v4(); // in working set
+    let uuid2 = Uuid::new_v4(); // in working set
+    let uuid3 = Uuid::new_v4(); // in working set but does not exist
+    let uuid4 = Uuid::new_v4(); // not in working set but exists
 
     {
         let mut txn = storage.txn()?;
         txn.add_to_working_set(uuid1)?;
         txn.add_to_working_set(uuid2)?;
-        txn.commit()?;
-    }
-
-    {
-        let mut txn = storage.txn()?;
+        txn.add_to_working_set(uuid3)?;
         assert!(txn.create_task(uuid1)?);
         txn.set_task(
             uuid1,
             taskmap_with(vec![("num".to_string(), "1".to_string())]),
         )?;
         assert!(txn.create_task(uuid2)?);
         txn.set_task(
             uuid2,
             taskmap_with(vec![("num".to_string(), "2".to_string())]),
         )?;
+        assert!(txn.create_task(uuid4)?);
+        txn.set_task(
+            uuid4,
+            taskmap_with(vec![("num".to_string(), "4".to_string())]),
+        )?;
         txn.commit()?;
     }
 
     {
         let mut txn = storage.txn()?;
         let mut tasks = txn.get_pending_tasks()?;
 
         let mut exp = vec![
```

Scandiravian replied:
That was entirely my fault. Your original comment did suggest to use `JOIN`, but I guess I was a bit tired when I wrote it and I didn't make sure I understood the code-base well enough

In any case; thank you for really thinking through the impact of my coding choices, so I can avoid introducing a bug that would frustrate your users :raised_hands:

djmitche reviewed (commented):
Really close! Just a few comments below.

Scandiravian committed:
use better variable name (all_tasks -> pending_tasks)

Scandiravian committed:
add comment explaining hard-to-read code

Scandiravian committed:
replicate and fix issue caused by `LEFT JOIN` instead of `JOIN`

Scandiravian commented:
Thank you so much for the review and for taking the time to thoroughly explain the bug that my choice of `LEFT JOIN` would have caused - it helps me improve as a developer

I've added a commit for each comment you had to make it easier to read through. I hope it's okay to run the changes by you one final time before I squash (actually `fixup`) everything into a single commit so it's ready to merge :crossed_fingers:

djmitche commented on taskchampion/src/storage/test.rs:
```diff
@@ -273,18 +273,17 @@ pub(super) fn delete_task_exists(mut storage: impl Storage) -> Result<()> {
+    let uuid1 = Uuid::new_v4(); // in working set
+    let uuid2 = Uuid::new_v4(); // in working set
+    let uuid3 = Uuid::new_v4(); // in working set but does not exist
+    let uuid4 = Uuid::new_v4(); // not in working set but exists= Uuid::new_v4();
```
Looks like a bit of extra text got included in that comment.

Scandiravian replied:
This is now fixed :smile:

djmitche reviewed (approved):
Looks great, thanks! And no problem at all making another round of review.

I can squash this when merging, so if you want to just make one more commit to fix the comment, that's all that should be necessary.

Commit:
fix typo

Scandiravian commented:
Hey @djmitche

If everything looks good to you, feel free to squash and merge the PR :smile:

djmitche reviewed (approved):
Sorry for the delay!

Scandiravian commented:
> Sorry for the delay!

No problem. I wasn't sure if I had communicated clearly that I thought I had solved all remaining issues 😅

Thanks so much for all your constructive feedback on my changes. This was an incredibly pleasant experience and I hope I'll get the chance to contribute to the project again
